module/input/Microphone/src/Microphone.cpp

```diff
@@ -0,0 +1,367 @@
+                std::string slot_value_str = std::string(slot_value->string);
+                            
+                intent->add_slot(slot_name_str, slot_value_str);
+                            
+                slots_element = slots_element->next;
+            }
+        }
+        
+        root_element = root_element->next;
+    }
+    free(root);
+    
+    return intent;
+}
+
+//TODO: there no way of testing this for the moment.
+//Use this reference example with "snd_pcm_poll_descriptors"
+//Which returns a file handle we can wait on.
+//https://gist.github.com/albanpeignier/104902
+/*
+bool start_microphone_capture() {
+    const char *device_name = "default";
+
+    int err = 0;
+    snd_pcm_t *capture_handle;
+    err = snd_pcm_open(&capture_handle, device_name, SND_PCM_STREAM_CAPTURE, 0);
+    if(err < 0) {
+        return false
+    }
+    
+    
+    
+    return true;
+}
+*/
+
+//handles: out parameter containing the stdin, stdout, stderr handles from the
+//spawned process
+//first char * in args array refers to the process name
+bool spawn_process(MicProcHandles *handles, char *const *args) {
+    bool success = false;
+    
+    pid_t pid = 0;
+    int stdout_pipe[2] = {};
+    int stderr_pipe[2] = {};
+    int stdin_pipe[2] = {};
+    posix_spawn_file_actions_t actions = {};
+    
+    char *path = getenv("PATH");
+    assert(path);
+    
+    size_t len = strlen(path) + 5 + 1;
+    char *env_path = (char*) malloc(len);
+    int written = sprintf(env_path, "PATH=%s", path);
+    assert(written > 0);
+    env_path[written] = 0;
+    
+    char *const envp[] = { 
+        (char*) "KALDI_DIR=", //KALDI_DIR env var must be blank!
+        (char*) env_path, 
+        NULL 
+    };
+    
+    if((pipe(stdout_pipe) == -1) || (pipe(stderr_pipe) == -1) || (pipe(stdin_pipe) == -1)) {
+        goto cleanup;
+    }
+    
+    posix_spawn_file_actions_init(&actions);
+    
+    posix_spawn_file_actions_addclose(&actions, stdout_pipe[0]);
+    posix_spawn_file_actions_addclose(&actions, stderr_pipe[0]);
+    posix_spawn_file_actions_addclose(&actions, stdin_pipe[1]);
+
+    posix_spawn_file_actions_adddup2(&actions, stdout_pipe[1], STDOUT_FILENO);
+    posix_spawn_file_actions_adddup2(&actions, stderr_pipe[1], STDERR_FILENO);
+    posix_spawn_file_actions_adddup2(&actions, stdin_pipe[0], STDIN_FILENO);
+    
+    posix_spawn_file_actions_addclose(&actions, stdout_pipe[1]);
+    posix_spawn_file_actions_addclose(&actions, stderr_pipe[1]);
+    posix_spawn_file_actions_addclose(&actions, stdin_pipe[0]);
+        
+    if(posix_spawnp(&pid, args[0], &actions, 0, args, envp) != 0) {
+        goto cleanup;
+    }
+    
+    success = true;
+cleanup:
+    if(stdout_pipe[1]) close(stdout_pipe[1]);
+    if(stderr_pipe[1]) close(stderr_pipe[1]);
+    if(stdin_pipe[0]) close(stdin_pipe[0]);
+
+    *handles = {stdout_pipe[0], stderr_pipe[0], stdin_pipe[1]};
+    
+    if(actions.__actions) {
```

Comment:
Isn't the double underscore a convention used to indicate that a field is private? Is this the right way of checking for whether the value is present?

Reply:
I added another variable "actions_inited" denoted whether that structure had been initialized. Now we check that variable instead.